apps/language_server/lib/language_server/providers/folding_range/token_pairs.ex

```diff
@@ -0,0 +1,97 @@
+defmodule ElixirLS.LanguageServer.Providers.FoldingRange.TokenPairs do
+  @moduledoc """
+  """
+
+  alias ElixirLS.LanguageServer.Providers.FoldingRange
+
+  @token_pairs %{
+    "(": [:")"],
+    "[": [:"]"],
+    "{": [:"}"],
+    # do blocks
+    do: [:block_identifier, :end],
+    block_identifier: [:block_identifier, :end],
+    # other special forms
+    with: [:do],
+    for: [:do],
+    case: [:do],
+    fn: [:end]
+  }
+
+  # Note
+  # This implementation allows for the possibility of 2 ranges with the same
+  # startLines but different endLines.
+  # It's not clear if that case is actually a problem.
+  @spec provide_ranges([FoldingRange.input()]) :: {:ok, [FoldingRange.t()]}
+  def provide_ranges(%{tokens: tokens}) do
+    ranges =
+      tokens
+      |> pair_tokens(@token_pairs)
+      |> convert_to_spec_ranges()
+
+    {:ok, ranges}
+  end
+
+  defp pair_tokens(tokens, kind_map) do
```

Comment:
Not sure why you take `kind_map` as an argument. I'd just use `@token_pairs` :thinking:

Reply:
Ah good catch! Yeah in an earlier iteration, I thought I needed to make multiple passes with different maps. This is no longer the case, so I should drop that.

Reply:
Fixed by: 2d20671